### ISSUE TITLE
Any way to provide custom config to ClickHouse?

### ISSUE THREAD

**Issue by julienma** [dokku/dokku-clickhouse#2]
Hey team,
I'm adapting the [docker-compose](https://github.com/plausible/hosting/blob/master/docker-compose.yml) file from [Plausible](https://plausible.io/docs/self-hosting) to make it work on dokku.

I'm using this dokku-clickhouse plugin, and everything works fine.

However, in the docker-compose file, they're adding 2 custom `.xlm` configuration files for Clickhouse, via docker's volume feature:

```
  plausible_events_db:
    image: yandex/clickhouse-server:21.3.2.5
    volumes:
      - event-data:/var/lib/clickhouse
      - ./clickhouse/clickhouse-config.xml:/etc/clickhouse-server/config.d/logging.xml:ro
      - ./clickhouse/clickhouse-user-config.xml:/etc/clickhouse-server/users.d/logging.xml:ro
```

Do you have any idea on how I could provide those 2 config files to the clickhouse container?

---

Here's what I tried:

1. it seems the only [volume mounted by the plugin](https://github.com/dokku/dokku-clickhouse/blob/master/functions#L79) is `-v "$SERVICE_HOST_ROOT/data:/var/lib/clickhouse"` (also see [clickhouse dockerfile](https://github.com/ClickHouse/ClickHouse/blob/master/docker/server/Dockerfile#L97)). I thought about copying the 2 custom `.xml` files in the `data` directory to make them available to the container, and then...
2. I thought I could use [`CLICKHOUSE_CUSTOM_ENV`](https://github.com/dokku/dokku-clickhouse/blob/96faa27d712b9c09e400988b51ebd42a8e5beb7c/functions#L60) to pass a config to clickhouse and tell it to load the custom configs from the `data` directory from step 1. But I can't find any ENV on clickhouse side that would let me pass a custom folder to load config from.

In fact it seems that clickhouse expects any custom configuration to be stored in 2 specific folders, `/etc/clickhouse-server/config.d/*.xml` and `/etc/clickhouse-server/usert.d/*.xml`, see https://github.com/ClickHouse/ClickHouse/tree/master/docker/server#volumes and https://clickhouse.tech/docs/en/operations/configuration-files/.

If there's no way currently to load / create on-the-fly a custom configuration, do you think it would be possible to add an option to this dokku plugin to mount 2 additional volumes for custom configuration files? E.g. add something along the lines of `-v "$SERVICE_HOST_ROOT/config/config.d/*.xml:/etc/clickhouse-server/config.d/*.xml" -v "$SERVICE_HOST_ROOT/config/users.d/*.xml:/etc/clickhouse-server/users.d/*.xml"`?

Thanks

**Comment by josegonzalez** [comment]
There isn't a way to do `*` for mounting afaik. We also currently don't have the ability to add custom docker options to services, which would enable this. The latter is something we'd like to do (there is an issue in the dokku-redis plugin for that).

Do you know if the docker image has any special way to specify extra configs for either of those things?

**Comment by julienma** [comment]
I'm not sure, their doc is not really crystal clear.

It seems possible to mount a [custom `config.xml`](https://github.com/ClickHouse/ClickHouse/tree/master/docker/server#start-server-instance-with-custom-configuration), e.g. `-v /path/to/your/config.xml:/etc/clickhouse-server/config.xml`. But that's probably a pain to have to copy the [WHOLE config](https://github.com/ClickHouse/ClickHouse/blob/master/programs/server/config.xml) just to customize a few lines, esp. since it changes quite often.

However—and if I understand correctly—there _might_ be a way to make customizations by mounting `.sh` scripts in `/docker-entrypoint-initdb.d/`, as these scripts would be run when service starts (see https://github.com/ClickHouse/ClickHouse/tree/master/docker/server#how-to-extend-this-image).

E.g. if dokku-clickhouse could mount smthg like `-v "$SERVICE_HOST_ROOT/config/docker-entrypoint-initdb.d/:/docker-entrypoint-initdb.d/"`, then I could probably write a script which would create my custom config file (`/etc/clickhouse-server/config.d/logging.xml`)?

**Comment by josegonzalez** [comment]
Yeah that might work. Does that directory have anything in it by default?

**Comment by julienma** [comment]
No, it's empty:

```sh
julien@b-dokku-test-20210412 ~  $ dokku clickhouse:enter plausible_events_db
-----> Filesystem changes may not persist after container restarts
root@2ce2b82a9bc6:/# la
bin  boot  dev  docker-entrypoint-initdb.d  .dockerenv  entrypoint.sh  etc  home  lib  lib32  lib64  libx32  media  mnt  opt  proc  root  run  sbin  srv  sys  tmp  usr  var
root@2ce2b82a9bc6:/# la docker-entrypoint-initdb.d/
root@2ce2b82a9bc6:/#
```

Also: https://github.com/ClickHouse/ClickHouse/blob/master/docker/server/Dockerfile#L87

**Comment by josegonzalez** [comment]
In that case, I think we can create an entrypoint directory in the service data dir and have it be mounted. A user can create the service, stop it, populate the directory with their scripts, then start again. Thoughts?

**Comment by julienma** [comment]
> A user can create the service, stop it, populate the directory with their scripts, then start again.

Is starting + stopping the service first required? Oh, because the data dir doesn't exist before that?

I never used the service data dir before, so don't have an opinion on where the entrypoint dir should live. But yeah, seems fine!

**Comment by josegonzalez** [comment]
It's not necessarily required, but service creation starts the service so you'd either need to have a way to not do that or pre-create the directories (which idk if we fill in not created ones).

**Comment by julienma** [comment]
As the service already mounts `-v "$SERVICE_HOST_ROOT/data:/var/lib/clickhouse"`, would that mean that the entrypoint dir will be available to the container from both `/docker-entrypoint-initdb.d/` and `/var/lib/clickhouse/docker-entrypoint-initdb.d/`? Would that be an issue?

**Comment by josegonzalez** [comment]
We'd want to create something like `$SERVICE_ROOT/entrypoint` and then have a mount flag like `-v "$SERVICE_HOST_ROOT/entrypoint:/docker-entrypoint-initdb.d"`.

We also should add this to the info output for the service so that end-users know what the path is that they can add entrypoint files to.

**Comment by julienma** [comment]
👍

**Comment by josegonzalez** [comment]
This was implemented in #14 and released in 0.7.0 of the plugin.
